From 4ecafa6a361f21aee197525d11f9622e02bd0d9e Mon Sep 17 00:00:00 2001
From: Rui Wang <ruwang@redhat.com>
Date: Thu, 21 Aug 2014 04:13:43 -0400
Subject: [edac] ghes: Update ghes error record info

Message-id: <1408594432-28006-9-git-send-email-ruwang@redhat.com>
Patchwork-id: 88495
O-Subject: [PATCH RHEL7.1 BZ1076655 08/17] EDAC, GHES: Update ghes error record info
Bugzilla: 1076655
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Aristeu Rozanski <aris@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Chen, Gong" <gong.chen@linux.intel.com>

Backport of:
Original commit: 56507694de3453076d73e0e9813349586ee67e59
Original Author: "Chen, Gong" <gong.chen@linux.intel.com>
Original Date: Fri, 18 Oct 2013 14:30:38 -0700

 In latest UEFI spec(by now it's 2.4) there are some new
 fields for memory error reporting. Add these new fields for
 ghes_edac interface.

Signed-off-by: Chen, Gong <gong.chen@linux.intel.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/edac/ghes_edac.c b/drivers/edac/ghes_edac.c
index 0ad797b..d5a98a4 100644
--- a/drivers/edac/ghes_edac.c
+++ b/drivers/edac/ghes_edac.c
@@ -314,6 +314,8 @@ void ghes_edac_report_mem_error(struct ghes *ghes, int sev,
   p += sprintf(p, "card:%d ", mem_err->card);
  if (mem_err->validation_bits & CPER_MEM_VALID_MODULE)
   p += sprintf(p, "module:%d ", mem_err->module);
+ if (mem_err->validation_bits & CPER_MEM_VALID_RANK_NUMBER)
+  p += sprintf(p, "rank:%d ", mem_err->rank);
  if (mem_err->validation_bits & CPER_MEM_VALID_BANK)
   p += sprintf(p, "bank:%d ", mem_err->bank);
  if (mem_err->validation_bits & CPER_MEM_VALID_ROW)
@@ -322,6 +324,15 @@ void ghes_edac_report_mem_error(struct ghes *ghes, int sev,
   p += sprintf(p, "col:%d ", mem_err->column);
  if (mem_err->validation_bits & CPER_MEM_VALID_BIT_POSITION)
   p += sprintf(p, "bit_pos:%d ", mem_err->bit_pos);
+ if (mem_err->validation_bits & CPER_MEM_VALID_MODULE_HANDLE) {
+  const char *bank = NULL, *device = NULL;
+  dmi_memdev_name(mem_err->mem_dev_handle, &bank, &device);
+  if (bank != NULL && device != NULL)
+   p += sprintf(p, "DIMM location:%s %s ", bank, device);
+  else
+   p += sprintf(p, "DIMM DMI handle: 0x%.4x ",
+         mem_err->mem_dev_handle);
+ }
  if (p > e->location)
   *(p - 1) = '\0';
 
diff --git a/include/linux/edac.h b/include/linux/edac.h
index 71d6441..bd50877 100644
--- a/include/linux/edac.h
+++ b/include/linux/edac.h
@@ -51,7 +51,7 @@ static inline void opstate_init(void)
 #define EDAC_MC_LABEL_LEN 31
 
 /* Maximum size of the location string */
-#define LOCATION_SIZE 80
+#define LOCATION_SIZE 256
 
 /* Defines the maximum number of labels that can be reported */
 #define EDAC_MAX_LABELS  8
-- 
1.7.1